### PR TITLE
ROX-13216: Resolver component: implement deployment resolution given deployment ID reference

### PR DESCRIPTION
**Pull request by fredrb**
## Description

Given a deployment ID provided by any of the handlers, build the deployment object using public interfaces from deployment, RBAC and Service stores. 

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- Unit tests added
- CI

### CODE REVIEW

File: sensor/kubernetes/eventpipeline/resolver/resolver_impl.go [hunk 1]
```diff
@@ -38,8 +49,47 @@ func (r *resolverImpl) runResolver() {
 
 // processMessage resolves the dependencies and forwards the message to the outputQueue
 func (r *resolverImpl) processMessage(msg *component.ResourceEvent) {
-	// TODO: resolve dependencies
+	if msg.DeploymentReference != nil {
+		referenceIds := msg.DeploymentReference(r.deploymentStore)
+
+		for _, id := range referenceIds {
+			preBuiltDeployment := r.deploymentStore.Get(id)
```
[the remark this answers is not in the document]
Comment (fredrb) [reply sensor/kubernetes/eventpipeline/resolver/resolver_impl.go:56]: Seems like the same type of race can happen elsewhere in the code. For example with the image scans:
https://github.com/stackrox/stackrox/blob/master/sensor/common/detector/detector.go#L483-L492

File: sensor/kubernetes/eventpipeline/resolver/resolver_test.go [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+	if deployment.GetId() != m.id {
+		m.error = fmt.Sprintf("IDs don't match: expected %s != %s", m.id, deployment.GetId())
+		return false
+	}
+
+	if deployment.GetServiceAccountPermissionLevel() != m.permissionLevel {
+		m.error = fmt.Sprintf("Permission level doesn't match %s != %s", m.permissionLevel, deployment.GetServiceAccountPermissionLevel())
+		return false
+	}
+
```
Comment (lvalerom) [inline sensor/kubernetes/eventpipeline/resolver/resolver_test.go:466]: I think the `deploymentMatcher` should also check the `Ports`

Comment (fredrb) [reply sensor/kubernetes/eventpipeline/resolver/resolver_test.go:516]: Done

**Review by lvalerom — APPROVED** [review]

File: sensor/kubernetes/eventpipeline/resolver/resolver_test.go [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+func (m *detectionObjectMatcher) Matches(target interface{}) bool {
+	event, ok := target.(*component.ResourceEvent)
+	if !ok {
+		m.error = "received message isn't a resource event"
+		return false
+	}
+
+	if !cmp.Equal(m.expected, event.CompatibilityDetectionDeployment) {
+		m.error = fmt.Sprintf("received detection deployment doesn't match expected: %s", cmp.Diff(m.expected, event.CompatibilityReprocessDeployments))
```
[the remark this answers is not in the document]
Comment (janisz) [reply sensor/kubernetes/eventpipeline/resolver/resolver_test.go:550]: - https://github.com/stackrox/stackrox/pull/11811

**Pull request merged**